geps/gep-1911.md

```diff
@@ -0,0 +1,238 @@
+# GEP-1911: Backend Protocol Selection
+
+* Issue: [#1911](https://github.com/kubernetes-sigs/gateway-api/issues/1911)
+* Status: Provisional
+
+(See status definitions [here](overview.md#status).)
+
+## TLDR
+
+Not all implementations support automatic protocol selection. Even in some cases protocols are disabled without an explicit opt-in (eg. websockets). Thus application developers need the ability to specify the protocol(s) that their application supports.
+
+## Goals
+
+- Support protocols that can have a Gateway '*Route' resource as a frontend
+- Consistent way of declaring backend protocols using a meta-resource
+- Support backends with multiple protocols on the same port (ie. h2c/http1)
```

Comment:
> Can you elaborate on your specific case ? do you need a backend that supports multiple protocols on the same port?

Yeah this is fairly common. For example I'm aware of users running TensorFlow Serving on CloudRun. It exposes HTTP and gRPC protocols for model prediction and it's being run on a single port that's multiplexing between them.

Reply:
Also K8s supports multiple Service `protocol` on the same port.

I added a section about this - please re-review.

Reply:
> Yeah this is fairly common. For example I'm aware of users running TensorFlow Serving on CloudRun. It exposes HTTP and gRPC protocols for model prediction and it's being run on a single port that's multiplexing between them.

I read a little bit about TensorFlow Service, it looks like it is two different ports. 8500 for GRPC and 8501 for HTTP.

The reason I am bringing this up is that I am trying to collect real uses cases for converting `AppProtocol` to a list.

Reply:
Also, it is still a goal on this GEP, not sure I found how this GEP suggests to support it?
few service protocols?

Reply:
> Also, it is still a goal on this GEP, not sure I found how this GEP suggests to support it?

It was buy not anymore. I'll remove it

Reply:
Actually, K8s supports TCP/UDP protocol on the same port - so I think this requirement stands.

Reply:
> The reason I am bringing this up is that I am trying to collect real uses cases for converting AppProtocol to a list.

An alternative to a list could be allowing multiple entries like how the mixed protocol KEP suggested it

https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1435-mixed-protocol-lb#proposal

Reply:
lets say we support a list, how would it help for your use case?

Reply:
Knative OSS and Cloud Run support the same portable API. It has a restrictions to listen only on one port since we need to capture request metrics to scale and we also enforce request concurrency.

So if there was a TensorFlow user which wanted to bring  that workload to Knative in order to get elastic autoscaling and scale to zero etc. they wouldn't be able to host both GRPC and HTTP using the same instance.

The TensorFlow Serving isn't a random use case - there are competing products doing the same thing - https://kserve.github.io

Reply:
We run Istio on CloudRun and do the same - gRPC and http on the same port